File: python/leetcode/symmetric-tree.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        right = root.right
        left = root.left

        a = self.is_same(right, left)

        return a

    def is_same(self, q, p):
        if p is None and q is None:
            return True

        if p is not None and q is not None:
            return (
                (p.val == q.val) and
                self.is_same(p.left, q.right) and
                self.is_same(p.right, q.left)
            )

        return False

File: typescript/leetcode/symmetric-tree.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}

function isSame(q: TreeNode | null, p: TreeNode | null) {
  if (!p && !q) {
      return true
  }

  if (p && q) {
      return p.val == q.val && isSame(p.left, q.right) && isSame(p.right, q.left)
  }

  return false
}

function isSymmetric(root: TreeNode | null): boolean {
  if (!root) {
    return false
  }

  return isSame(root.right, root.left)
};
